linera-service/src/util.rs

```diff
@@ -1,132 +1,80 @@
+    // (same thing for target/release).
     {
-        features.push("scylladb");
+        let components = current_binary_path
+            .components()
+            .map(Component::as_os_str)
+            .collect::<Vec<_>>();
+        let len = components.len();
+        if len >= 3
+            && components[len - 3] == "target"
+            && (components[len - 2] == "debug" || components[len - 2] == "release")
+            && components[len - 1] == "deps"
+        {
+            current_binary_path.pop();
+        }
     }
-    #[cfg(feature = "aws")]
-    {
-        features.push("aws");
+
+    let path = current_binary_path.join(name);
+    let version = env!("CARGO_PKG_VERSION");
+    if !path.exists() {
+        error!(
+            "Cannot find a binary {name} in the directory {}. \
+             Consider using `cargo install {package}` or `cargo build -p {package}`",
+            current_binary_path.display()
+        );
+        bail!("Failed to resolve binary {name}");
     }
-    features
-}
 
-async fn cargo_force_build_binary(name: &'static str, package: Option<&'static str>) -> PathBuf {
-    let package = package.unwrap_or(env!("CARGO_PKG_NAME"));
-    let mut build_command = Command::new("cargo");
-    build_command.args(["build", "-p", package]);
-    let is_release = if let Ok(var) = env::var(CARGO_ENV) {
-        let extra_args = var.split_whitespace();
-        build_command.args(extra_args.clone());
-        let extra_args: HashSet<_> = extra_args.into_iter().map(str::trim).collect();
-        extra_args.contains("-r") || extra_args.contains("--release")
-    } else {
-        false
-    };
-    // Use the same features as the current environment so that we don't rebuild as often.
-    let features = detect_current_features().join(",");
-    build_command
-        .arg("--no-default-features")
-        .arg("--features")
-        .arg(features);
-    build_command.args(["--bin", name]);
-    info!("Running compiler: {:?}", build_command);
-    assert!(build_command
-        .spawn()
-        .unwrap()
-        .wait()
+    // Quick version check.
+    let version_message = Command::new(&path)
+        .arg("--version")
+        .output()
         .await
-        .unwrap()
-        .success());
-    let mut cargo_locate_command = Command::new("cargo");
-    cargo_locate_command.args(["locate-project", "--workspace", "--message-format", "plain"]);
-    let output = cargo_locate_command.output().await.unwrap().stdout;
-    let workspace_path = Path::new(std::str::from_utf8(&output).unwrap().trim())
-        .parent()
-        .unwrap();
-    if is_release {
-        workspace_path
-            .join("target/release")
-            .join(name)
-            .canonicalize()
-            .unwrap()
-    } else {
-        workspace_path
-            .join("target/debug")
-            .join(name)
-            .canonicalize()
-            .unwrap()
+        .map_err(|e| {
```

Comment:
You can just use `with_context` here (since formatting is expensive and you want to do it lazily). You don't need the `map_err` as the error will be shown anyway:


```
Error: Failed to execute and retrieve version from the binary {name} in directory {}

Caused by:
    {e}
```

Reply:
`with_context` seems to only work if you have an `anyhow::Result` already.

Reply:
The documentation says the error type just has to be `StdError + Send + Sync + 'static`.